keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1430 @@
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Today, the scheduler surfaces a specific pod condition: `Scheduled` that clearly
+identifies whether a pod got scheduled by the scheduler and when scheduling
+completed. However, no specific conditions around initialization of successfully
+scheduled pods from the perspective of completion of pod sandbox creation is
+surfaced to cluster administrators in a scoped and consumable fashion.
+
+There is an existing pod condition: `Initialized` that tracks execution of init
+containers. For pods without init containers, the `Initialized` condition is set
+when the Kubelet starts to process a pod before any sandbox creation activities
+start. For pods with init containers, the `Initialized` condition is set when
+init containers have been pulled and executed to completion. Therefore, the
+existing `Initialized` condition is insufficient and inaccurate for tracking
+completion of sandbox creation of all pods in a cluster. This distinction
+becomes especially relevant in multi-tenant clusters where individual tenants
+own the pod specs (including the set of init containers) while the cluster
+administrators are in charge of storage plugins, networking plugins and
+container runtime handlers.
+
+A dedicated condition around readiness of the pod sandbox will benefit cluster
+operators (especially of multi-tenant clusters) who are responsible for
+configuration and operational aspects of the various components that play a role
+in pod sandbox creation: CSI plugins, CRI runtime and associated runtime
+handlers, CNI plugins, etc. The duration between `lastTransitionTime` field of
+the `SandboxReady` condition and the existing `Scheduled` condition will allow
+metrics collection services to surface total latency of all the components
+involved in pod sandbox creation as an SLI. Cluster operators can use this to
+publish SLOs around pod initialization to their customers who launch workloads
+on the cluster.
+
+Custom pod controllers/operators can use a dedicated condition indicating
+completion of pod sandbox creation to make better decisions around how to
+reconcile a pod failing to become ready. As a specific example, a custom
+controller for managing pods that refer to PVCs associated with node local
+storage (e.g. Rook-Ceph) may decide to recreate PVCs (based on a specified PVC
+template) if the sandbox creation is repeatedly failing to complete. Such a
+controller can leave PVCs intact and only recreate pods if sandbox creation
+completes successfully but the pod's containers fail to become ready.
+
+When a pod is gracefully terminated, the status of `SandboxReady` condition will
+be set to `false`. The duration between a pod's `DeletionTimeStamp` and
+`lastTransitionTime` of `SandboxReady` condition will indicate the latency of
```

Comment:
For my own education, does the duration between a pod's `DeletionTimeStamp` and
`lastTransitionTime` of `Ready` condition serve the same purpose?

Reply:
This is a good point - it will depend on the readiness probes configured on the pod. If one of the readiness probes reports a False condition well before containers being probed are actually terminated, the `Ready` condition will be set to False in an earlier call to `generateAPIPodStatus` => `GeneratePodReadyCondition`. In such cases, the duration between a pod's `DeletionTimeStamp` and `lastTransitionTime` of `Ready` condition will _not_ be the same as the duration between pod's `DeletionTimeStamp` and `lastTransitionTime` of `SandboxReady` condition (as the latter's transitions will be based on state of the CRI sandbox).

Reply:
uh makes sense. I forgot about Readiness probe in this case.